compiler/rustc_mir_transform/src/coverage/mod.rs

```diff
@@ -295,6 +295,37 @@ fn make_code_region(
     })
 }
 
+/// If `llvm-cov` sees a code region that is improperly ordered (end < start),
+/// it will immediately exit with a fatal error. To prevent that from happening,
+/// discard regions that are improperly ordered, or might be interpreted in a
+/// way that makes them improperly ordered.
+fn check_code_region(code_region: CodeRegion) -> Option<CodeRegion> {
+    let CodeRegion { file_name: _, start_line, start_col, end_line, end_col } = code_region;
+
+    // Line/column coordinates are supposed to be 1-based. If we ever emit
+    // coordinates of 0, `llvm-cov` might misinterpret them.
+    let all_nonzero = [start_line, start_col, end_line, end_col].into_iter().all(|x| x != 0);
+    // Coverage mappings use the high bit of `end_col` to indicate that a
+    // region is actually a "gap" region, so make sure it's unset.
+    let end_col_has_high_bit_unset = (end_col & (1 << 31)) == 0;
+    // If a region is improperly ordered (end < start), `llvm-cov` will exit
+    // with a fatal error, which is inconvenient for users and hard to debug.
+    let is_ordered = (start_line, start_col) <= (end_line, end_col);
+
+    if all_nonzero && end_col_has_high_bit_unset && is_ordered {
+        Some(code_region)
+    } else {
+        debug!(
```

Comment:
It sounds like you originally had `debug_assert!()`s in places like this? I'm curious why you decided against leaving those in 🙂

Reply:
I've been going back and forth because I'm genuinely unsure whether they should be there or not. 🤷‍♂️

I took them out partly because it felt “messy” to have so much duplicated logic/diagnostics between the asserts, debug logs, and returns. But thinking about it once more, I'm coming back around to the idea of having a single `debug_assert!` right at the end. That minimizes mess, while still giving the benefit of making noise when this happens during development.

Reply:
@wesleywiser please review